### ISSUE TITLE
Making a Go program 42% faster with a one character change

### ISSUE THREAD

**Issue by luantranminh** [luantranminh/til#145]
https://hmarr.com/blog/go-allocation-hunting/
TL;DR: 
- compiler' technique called **escape analysis**: figure out when some memory needs to live on the heap
- Inside loop, we copy the heap allocated out of the slice onto the stack, then by returning *pointer* we create an (escaping) pointer to the copy of the struct. -> We just need to move the ampersand up a bit so we take a reference to the struct in the slice rather than copying it
```go
// before
type Ruleset []Rule

func (r Ruleset) Match(path string) (*Rule, error) {
  for i := len(r) - 1; i >= 0; i-- {
    rule := r[i]
    match, err := rule.Match(path)
    if match || err != nil {
      return &rule, err
    }
  }
  return nil, nil
}

// after
 func (r Ruleset) Match(path string) (*Rule, error) {
 	for i := len(r) - 1; i >= 0; i-- {
-		rule := r[i]
+		rule := &r[i]
 		match, err := rule.Match(path)
 		if match || err != nil {
-			return &rule, err
+			return rule, err
 		}
 	}
 	return nil, nil
 }
```

#### Remarkable used tools:
- [hyperfine](https://github.com/sharkdp/hyperfine): A command-line benchmarking tool.
     - Statistical analysis across multiple runs
     - Cache-clearing commands can be set up before each timing run.
- Showing complier process: `go build -gcflags=-m *.go 2>&1 | grep codeowners.go`
- Go's built-in profile tool:  
     - interactive profile visualization tool : go tool pprof -http=":8000" ./codeowners ./cpu.pprof
     - pprof snippet
```go
pprofFile, pprofErr := os.Create("cpu.pprof")
if pprofErr != nil {
  log.Fatal(pprofErr)
}
pprof.StartCPUProfile(pprofFile)
defer pprof.StopCPUProfile()
```
